core/spec/models/spree/concerns/user_address_book_spec.rb

```diff
@@ -289,82 +371,82 @@ module Spree
           user.persist_order_address(order)
+
+          expect( user.bill_address_id ).to eq order.bill_address_id
+          expect( user.ship_address_id ).to eq order.ship_address_id
+          expect( user.bill_address_id ).not_to eq user.ship_address_id
+
+          expect( user.bill_address).to eq order.bill_address
+          expect( user.ship_address).to eq order.ship_address
         end
       end
 
       context "when automatic_default_address preference is false" do
+        let(:order) { build :order }
+
         before do
           stub_spree_preferences(automatic_default_address: false)
-          expect(user).to receive(:save_in_address_book).with(kind_of(Hash), false).twice
-          # and not the optional 2nd argument
         end
 
-        it "does not set the default: true flag" do
-          order = build(:order)
+        it "will save only the default ship address on user as it is the first address at all" do
           user.persist_order_address(order)
+
+          expect( user.bill_address_id ).to eq order.bill_address_id
+          expect( user.ship_address_id ).to eq order.ship_address_id
+
+          expect( user.bill_address).to be_nil
+          expect( user.ship_address).to eq order.ship_address
         end
       end
 
-      context "when address is nil" do
+      context "when one order address is nil" do
         context "when automatic_default_address preference is at a default of true" do
           before do
             stub_spree_preferences(automatic_default_address: true)
-            expect(user).to receive(:save_in_address_book).with(kind_of(Hash), true).once
           end
 
           it "does not call save_in_address_book on ship address" do
             order = build(:order)
             order.ship_address = nil
 
+            expect(user).to receive(:save_in_address_book).with(kind_of(Hash), true, :billing).once
             user.persist_order_address(order)
           end
 
           it "does not call save_in_address_book on bill address" do
             order = build(:order)
             order.bill_address = nil
 
+            expect(user).to receive(:save_in_address_book).with(kind_of(Hash), true).once
             user.persist_order_address(order)
           end
         end
-      end
 
-      context "when automatic_default_address preference is false" do
-        before do
-          stub_spree_preferences(automatic_default_address: false)
-          expect(user).to receive(:save_in_address_book).with(kind_of(Hash), false).once
-        end
+        context "when automatic_default_address preference is false" do
+          before do
+            stub_spree_preferences(automatic_default_address: false)
+          end
 
-        it "does not call save_in_address_book on ship address" do
-          order = build(:order)
-          order.ship_address = nil
+          it "does not call save_in_address_book on ship address" do
+            order = build(:order)
```

Comment:
I think this line (which is repeated in the next test) may be extracted in a `let`

Reply:
Done